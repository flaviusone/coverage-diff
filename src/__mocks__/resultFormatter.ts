export const resultFormatter = () => 'resultFormatter mock';
