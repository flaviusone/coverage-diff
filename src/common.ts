/* eslint-disable no-undef */
export interface ICoverageSummary {
  lines: ICoverageInfo;
  statements: ICoverageInfo;
  functions: ICoverageInfo;
  branches: ICoverageInfo;
}

export interface ICoverageInfo {
  total: number;
  covered: number;
  skipped: number;
  pct: number; // Coverage percentage (covered/total ratio).
}

export interface IJsonSummary {
  [key: string]: ICoverageSummary;
}

export interface ICoverageDiffOutput {
  diff: IJsonSummary;
  results: string;
  regression: boolean;
}

export type Criteria = 'lines' | 'branches' | 'functions' | 'statements';

export interface IConfigOptions {
  checkCriteria?: Array<Criteria>;
  /* Fail coverage check if per-file coverage is lower */
  coverageThreshold?: number;
  /* Fail coverage check if per-file coverage decrease is lower */
  coverageDecreaseThreshold?: number;
}

export interface IDiffCheckResults {
  files: IFilesResults;
  totals: IFileResultFormat;
  diff: IJsonSummary;
  regression: boolean;
}

export interface IFilesResults {
  [key: string]: IFileResultFormat;
}

export interface IFileResultFormat {
  deltas: IFileResultFields;
  pcts: IFileResultFields;
  decreased: boolean;
}

export interface IFileResultFields {
  lines: number;
  functions: number;
  statements: number;
  branches: number;
}
