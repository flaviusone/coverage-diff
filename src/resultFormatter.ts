import markdownTable from 'markdown-table';
import { getSummaryPercentages } from './helpers';
import { IFilesResults, ICoverageSummary } from './common';

export const resultFormatter = (
  files: IFilesResults,
  total: ICoverageSummary
): string => {
  const formattedFiles = formatFilesResults(files);
  const formattedTotal = formatTotal(total);
  return `${formattedFiles}${formattedTotal}`;
};

const formatTotal = (total: ICoverageSummary): string => {
  const table: Array<(string | number)[]> = [];
  const { lines, branches, functions, statements } = getSummaryPercentages(
    total
  );
  table.push(['Lines', 'Branches', 'Functions', 'Statements']);
  table.push([`${lines}%`, `${branches}%`, `${functions}%`, `${statements}%`]);

  return '\n\nTotal:\n\n' + markdownTable(table);
};

const formatFilesResults = (files: IFilesResults): string => {
  let noChange = true;
  const table: Array<(string | number)[]> = [];
  const header = [
    'Ok',
    'File',
    'LinesÎ”',
    'BranchesÎ”',
    'FunctionsÎ”',
    'StatementsÎ”'
  ];
  table.push(header);

  Object.keys(files).forEach(file => {
    const { deltas, decreased } = files[file];
    const row = [
      decreased ? 'ðŸ”´' : 'âœ…',
      file,
      formatDelta(deltas.lines),
      formatDelta(deltas.branches),
      formatDelta(deltas.functions),
      formatDelta(deltas.statements)
    ];

    table.push(row);
    noChange = false;
  });

  return noChange ? 'Coverage values did not changeðŸ‘Œ.' : markdownTable(table);
};

const formatDelta = (num: number): string => {
  return num >= 0 ? `+${num}%` : `${num}%`;
};
