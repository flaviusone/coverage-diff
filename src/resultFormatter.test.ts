import { FilesResults, FileResultFormat } from './common';
import { resultFormatter } from './resultFormatter';

const filesResults: FilesResults = {
  file1: {
    isNewFile: false,
    deltas: {
      lines: 10,
      statements: -10,
      functions: 20,
      branches: -30
    },
    pcts: {
      lines: 80,
      statements: 20,
      functions: 3,
      branches: 14
    },
    decreased: true,
    belowThreshold: false
  },
  file2: {
    isNewFile: false,
    deltas: {
      lines: 10,
      statements: 10,
      functions: 20,
      branches: 30
    },
    pcts: {
      lines: 20,
      statements: 5,
      functions: 2,
      branches: 8
    },
    decreased: false,
    belowThreshold: false
  },
  file3: {
    isNewFile: false,
    deltas: {
      lines: 10,
      statements: 10,
      functions: 20,
      branches: 30
    },
    pcts: {
      lines: 20,
      statements: 5,
      functions: 2,
      branches: 8
    },
    decreased: false,
    belowThreshold: true
  },
  file4: {
    isNewFile: true,
    deltas: {
      lines: 10,
      statements: 10,
      functions: 20,
      branches: 30
    },
    pcts: {
      lines: 20,
      statements: 5,
      functions: 2,
      branches: 8
    },
    decreased: false,
    belowThreshold: false
  },
  file5: {
    isNewFile: true,
    deltas: {
      lines: 10,
      statements: 10,
      functions: 20,
      branches: 30
    },
    pcts: {
      lines: 20,
      statements: 5,
      functions: 2,
      branches: 8
    },
    decreased: false,
    belowThreshold: true
  }
};

const totalResults: FileResultFormat = {
  isNewFile: false,
  deltas: {
    lines: 100,
    functions: 100,
    branches: 100,
    statements: 100
  },
  pcts: {
    lines: 100,
    statements: 100,
    functions: 100,
    branches: 100
  },
  decreased: false,
  belowThreshold: false
};

describe('resultFormatter', () => {
  it('should format files results as markdown table', () => {
    expect(
      resultFormatter(filesResults, totalResults, 1, 2, 3)
    ).toMatchSnapshot();
  });

  it('should use defaults', () => {
    expect(resultFormatter(filesResults, totalResults)).toMatchSnapshot();
  });

  it('should default to coverageThreshold for newFileCoverageThreshold', () => {
    expect(resultFormatter(filesResults, totalResults, 1, 2)).toMatchSnapshot();
  });

  it('should format files results as markdown table with defaults', () => {
    expect(resultFormatter(filesResults, totalResults)).toMatchSnapshot();
  });

  it("should print descriptive message if coverage did't change", () => {
    expect(resultFormatter({}, totalResults)).toMatchSnapshot();
  });
});
