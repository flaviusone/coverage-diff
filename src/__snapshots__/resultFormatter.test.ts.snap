// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resultFormatter should format files results as markdown table 1`] = `
"| Ok  | File  | LinesÎ” | BranchesÎ” | FunctionsÎ” | StatementsÎ” |
| --- | ----- | ------ | --------- | ---------- | ----------- |
| ðŸ”´  | file1 | +10%   | -30%      | +20%       | -10%        |
| âœ…   | file2 | +10%   | +30%      | +20%       | +10%        |

Total:

| Lines | Branches | Functions | Statements |
| ----- | -------- | --------- | ---------- |
| 100%  | 100%     | 100%      | 100%       |"
`;

exports[`resultFormatter should print descriptive message if coverage did't change 1`] = `
"Coverage values did not changeðŸ‘Œ.

Total:

| Lines | Branches | Functions | Statements |
| ----- | -------- | --------- | ---------- |
| 100%  | 100%     | 100%      | 100%       |"
`;
